soco/music_services/music_service.py

```diff
@@ -0,0 +1,780 @@
+
+
+# pylint: disable=too-many-instance-attributes, protected-access
+class MusicServiceSoapClient(object):
+
+    """A SOAP client for accessing Music Services
+
+    This class handles all the necessary authentication for accessing third
+    party music services. You are unlikely to need to use it yourself.
+
+    """
+
+    def __init__(self, endpoint, timeout, music_service):
+        """ Constructor
+
+        Args:
+             endpoint (str): The SOAP endpoint. A url.
+             timeout (int): Timeout the connection after this number of
+                 seconds
+             music_service (MusicService): The MusicService object to which
+                 this client belongs
+
+        """
+
+        self.endpoint = endpoint
+        self.timeout = timeout
+        self.music_service = music_service
+        self.namespace = 'http://www.sonos.com/Services/1.1'
+
+        # Spotify uses gzip. Others may do so as well. Unzipping is handled
+        # for us by the requests library. We need to set the user agent to
+        # a genuine Sonos value. Google Play music seems to want this. The
+        # firmware release number (after 'Sonos/" is obviously fake, and
+        # higher than current values, in case Google only offers certain
+        # services to certain firmware rel eases. ") #
+
+        self.http_headers = {
+            'Accept-Encoding': 'gzip, deflate',
+            # NB Google Play seems to be sensitive to the User-Agent string
+            # This one works, but similar ones don't
+            'User-agent': 'Linux UPnP/1.0 Sonos/99'
+        }
+        self._device = SoCo.any_soco()
+        self._device_id = self._device.systemProperties.GetString(
+            [('VariableName', 'R_TrialZPSerial')])['StringValue']
+
+    def get_soap_header(self):
+        """ Generate the SOAP authentication header for the related service.
+
+        This header contains all the necessary authentication details.
+
+        Returns:
+            (str): A string representation of the XML content of the SOAP
+                header
+
+        """
+
+        # According to the SONOS SMAPI, this header must be sent with all SOAP
+        # requests.
+        music_service = self.music_service
+        credentials_header = XML.Element(
+            "credentials", {'xmlns': "http://www.sonos.com/Services/1.1"})
+        device_id = XML.SubElement(credentials_header, 'deviceId')
+        device_id.text = self._device_id
+        device_provider = XML.SubElement(credentials_header, 'deviceProvider')
+        device_provider.text = 'Sonos'
+        if music_service.account.oa_device_id:
+            # OAuth account credentials are present. We must use them to
+            # authenticate.
+            login_token = XML.Element('loginToken')
+            token = XML.SubElement(login_token, 'token')
+            token.text = music_service.account.oa_device_id
+            key = XML.SubElement(login_token, 'key')
```

Comment:
I must admit that everytime I look at XML generation code I feel torn. On the one hand, I really think we should use the library to create them, instead of doing by hand, to get rid of excaping issues etc. On the other hand it always annoys me, that it gets to long winded. Every time we add a subelement is two lines and a use-once-and-throw-away variable. I sometimes think that maybe it would be better, to do something like:
```python
XML.SubElement(login_token, 'token').text = music_service.account.oa_device_id
```
although, it is a bit harder to read. I don't know, as I said. Torn.